### PR TITLE
[RFC] Add per-service-level timeouts

### PR DESCRIPTION
**Pull request by psarna**
Ref: #7617

This überseries adds timeout parameters to service levels. It's RFC mainly because it cannot be merged on its own, as it's based on top of not-yet-merged service level series: #7867.
After the aforementioned series is merged, this one will be rebased on top of master.
Moreover, here's the (ever-growing) TODO list:
 [ ] More docs paragraphs should be added to `docs/service_levels.md`.
 [ ] More tests should be added, especially wrt. multiple service levels applicable for a user (since roles have a hierarchy) 
 [ ] currently, if multiple service levels apply to a role, the algorithm for picking the applicable parameters simply chooses a service level which comes first lexicographically. While deterministic, this heuristics is also ridiculously useless. Some more sensible ideas include picking the lowest timeout values and returning such a combination or choosing a service level which is the most specific - (e.g. when a user has role A which inherits from roles B and C, role A is "closest" to the user, so its service level options should apply).

Per-service-level timeouts can be set up in the form of service level parameters, which can in turn be attached to roles. Setting up and modifying role-specific timeouts can be achieved like this:
```cql
CREATE SERVICE LEVEL sl2 WITH read_timeout = 500ms AND write_timeout = 200ms AND cas_timeout = 2s;
ATTACH SERVICE LEVEL sl2 TO cassandra;
ALTER SERVICE LEVEL sl2 WITH write_timeout = 'null';
```

### CODE REVIEW

**Comment by psarna** [comment]
This series is still on top of unmerged #7867, but now it's more thoroughly tested, so we can progress with #7867.

**Comment by psarna** [comment]
Rebased, one newly introduced test still fails, probably due to a bug in the test itself. Still investigating

**Comment by psarna** [comment]
v2:
 - fixed the test issue - it was a faulty test, what was needed to be added was a memtable flush to ensure disk access for queries
 - added more checks to the test case, including a more complex tree of granted roles
 - added a paragraph in docs/service_levels.md, explaining the idea

**Comment by psarna** [comment]
rebased

**Comment by psarna** [comment]
I added the docs entry, which closes my original TODO list and makes this series finally ready. Reviewers welcome.

**Comment by avikivity** [comment]
Reviewing.

File: cql3/Cql.g [hunk 1]
```diff
@@ -1621,6 +1621,7 @@ property[cql3::statements::property_definitions& props]
 propertyValue returns [sstring str]
     : c=constant           { $str = c->get_raw_text(); }
     | u=unreserved_keyword { $str = u; }
+    | K_NULL { $str = "null"; }
```
Comment (avikivity) [inline cql3/Cql.g:1624]: What's the difference between x = NULL and x = 'null'? Won't they return the same thing?

Maybe propertyValue should return std::optional<sstring> to indicate null.

File: service/migration_manager.hh [hunk 2]
```diff
@@ -121,7 +121,7 @@ public:
 
     future<> announce_new_keyspace(lw_shared_ptr<keyspace_metadata> ksm, api::timestamp_type timestamp);
 
-    future<> announce_column_family_update(schema_ptr cfm, bool from_thrift, std::vector<view_ptr>&& view_updates);
+    future<> announce_column_family_update(schema_ptr cfm, bool from_thrift, std::vector<view_ptr>&& view_updates, std::optional<api::timestamp_type> timestamp);
```
Comment (avikivity) [inline service/migration_manager.hh:124]: Please explain (in a comment here and in the commit changelog) the role of the timestamp. It's not the schema timestamp, so what is it used for?

File: db/system_distributed_keyspace.cc [hunk 3]
```diff
@@ -160,6 +161,33 @@ system_distributed_keyspace::system_distributed_keyspace(cql3::query_processor&
...
+        for (const std::string_view& col_name : timeout_columns) {
+            bytes options_name = to_bytes(col_name.data());
+            if (schema->get_column_definition(options_name)) {
+                continue;
+            }
+            updated = true;
+            b.with_column(options_name, duration_type, column_kind::regular_column);
+        }
+        if (!updated) {
+            return make_ready_future<>();
```
Comment (avikivity) [inline db/system_distributed_keyspace.cc:181]: I feel like this should be much simpler, like `co_await execute_cql("ALTER TABLE ... ADD COLUMN ...");`

Comment (avikivity) [reply db/system_distributed_keyspace.cc:181]: (it can't be done for the bootstrap tables used for schemas, but for these higher level tables, I don't see why we need to C++ it)

File: db/system_distributed_keyspace.cc [hunk 4]
```diff
@@ -596,8 +596,23 @@ future<qos::service_levels_info> system_distributed_keyspace::get_service_level(
...
+    co_await _qp.execute_internal(format("UPDATE {}.{} SET read_timeout = ? "
+                ", write_timeout = ?, range_read_timeout = ?, counter_write_timeout = ? "
+                ", truncate_timeout = ?, cas_timeout = ?, other_timeout = ? "
+                "WHERE service_level = ?;", NAME, SERVICE_LEVELS),
+                db::consistency_level::ONE,
+                internal_distributed_query_state(),
+                {to_data_value(slo.read_timeout),  to_data_value(slo.write_timeout),  to_data_value(slo.range_read_timeout),
+                to_data_value(slo.counter_write_timeout),  to_data_value(slo.truncate_timeout),
+                to_data_value(slo.cas_timeout),  to_data_value(slo.other_timeout), service_level_name});
+    co_return;
```
Comment (avikivity) [inline db/system_distributed_keyspace.cc:639]: co_note: co_return isn't needed (but does no harm)

File: cql3/statements/sl_prop_defs.cc [hunk 5]
```diff
@@ -33,7 +34,7 @@ void sl_prop_defs::validate() {
         "read_timeout", "write_timeout", "range_read_timeout", "counter_write_timeout", "truncate_timeout", "cas_timeout", "other_timeout"
     };
     auto get_duration = [&] (const std::optional<sstring>& repr) -> std::optional<lowres_clock::duration> {
-        if (!repr) {
+        if (!repr || boost::algorithm::iequals(*repr, "null")) {
```
Comment (avikivity) [inline cql3/statements/sl_prop_defs.cc]: As commented before, this conflicts with a literal string saying null.

File: db/system_distributed_keyspace.cc [hunk 6]
```diff
@@ -567,14 +567,30 @@ system_distributed_keyspace::get_cdc_desc_v1_timestamps(context ctx) {
...
     return _qp.execute_internal(prepared_query, db::consistency_level::ONE, internal_distributed_query_state(), {}).then([] (shared_ptr<cql3::untyped_result_set> result_set) {
         qos::service_levels_info service_levels;
         for (auto &&row : *result_set) {
             auto service_level_name = row.get_as<sstring>("service_level");
-            qos::service_level_options slo{};
+            qos::service_level_options slo{
+                .read_timeout = get_duration(row, "read_timeout"),
+                .write_timeout = get_duration(row, "write_timeout"),
+                .range_read_timeout = get_duration(row, "range_read_timeout"),
+                .counter_write_timeout = get_duration(row, "reacounter_write_timeoutd_timeout"),
```
Comment (avikivity) [inline db/system_distributed_keyspace.cc:589]: The literal string in the end of the line got mangled.

File: db/system_distributed_keyspace.cc [hunk 7]
```diff
@@ -588,7 +604,15 @@ future<qos::service_levels_info> system_distributed_keyspace::get_service_level(
         if (!result_set->empty()) {
             auto &&row = result_set->one();
             auto service_level_name = row.get_as<sstring>("service_level");
-            qos::service_level_options slo{};
+            qos::service_level_options slo{
+                .read_timeout = get_duration(row, "read_timeout"),
+                .write_timeout = get_duration(row, "write_timeout"),
+                .range_read_timeout = get_duration(row, "range_read_timeout"),
+                .counter_write_timeout = get_duration(row, "reacounter_write_timeoutd_timeout"),
```
Comment (avikivity) [inline db/system_distributed_keyspace.cc:611]: Here too.

File: cql3/statements/list_service_level_statement.cc [hunk 8]
```diff
@@ -58,7 +58,15 @@ list_service_level_statement::execute(query_processor& qp,
...
 
-    static thread_local const std::vector<lw_shared_ptr<column_specification>> metadata({make_column("service_level", utf8_type)});
+    static thread_local const std::vector<lw_shared_ptr<column_specification>> metadata({make_column("service_level", utf8_type),
+        make_column("read_timeout", duration_type),
+        make_column("write_timeout", duration_type),
+        make_column("range_read_timeout", duration_type),
+        make_column("counter_write_timeout", duration_type),
+        make_column("truncate_timeout", duration_type),
+        make_column("cas_timeout", duration_type),
+        make_column("other_timeout", duration_type),
```
Comment (avikivity) [inline cql3/statements/list_service_level_statement.cc:68]: Having different timeouts for different operation types is dubious and I wish Cassandra hadn't done it.

Maybe we can live with just one timeout per session.

File: service/client_state.hh [hunk 9]
```diff
@@ -174,11 +176,12 @@ public:
         _driver_version = std::move(driver_version);
     }
 
-    client_state(external_tag, auth::service& auth_service, timeout_config timeout_config, const socket_address& remote_address = socket_address(), bool thrift = false)
+    client_state(external_tag, auth::service& auth_service, qos::service_level_controller* sl_controller, timeout_config timeout_config, const socket_address& remote_address = socket_address(), bool thrift = false)
```
Comment (avikivity) [inline service/client_state.hh:179]: Do we need the controller here, or just out service level?

File: cql3/statements/list_service_level_statement.cc (hunk 8 above)
Comment (psarna) [reply cql3/statements/list_service_level_statement.cc:68]: I'd actually be very happy to rework this series to provide just a single timeout. This explosion is terrible and I only did it because we already have in scylla.yaml, but having just one per-session timeout would make this series so much simpler.

I'll go ahead and do it in v2 and see how it goes.

File: db/system_distributed_keyspace.cc (hunk 6 above)
Comment (psarna) [reply db/system_distributed_keyspace.cc:589]: Ugh. Now I like the idea of unifying all these timeouts into a single per-service-level timeout even more.

File: service/client_state.cc [hunk 10]
```diff
@@ -270,3 +271,22 @@ future<> service::client_state::ensure_exists(const auth::resource& r) const {
...
+
+future<> service::client_state::maybe_update_per_service_level_params() {
+    if (_sl_controller && _user && _user->name) {
+        const auto& role_manager = _auth_service->underlying_role_manager();
+        auto role_set = co_await role_manager.query_granted(_user->name.value(), auth::recursive_role_query::yes);
+        auto slo_opt = co_await _sl_controller->find_service_level(role_set);
+        if (!slo_opt) {
+            co_return;
+        }
+        const qos::service_level_options& slo = *slo_opt;
```
Comment (avikivity) [inline service/client_state.cc:283]: I see why we need _sl_controller now. Not too happy about it, but it can remain.

Maybe _sl_controller should keep track of its service levels and propagate changes to them, instead of the client states polling for changes.

File: cql3/Cql.g (hunk 1 above)
Comment (psarna) [reply cql3/Cql.g:1624]: Yeah, but that change would be rather intrusive... It isn't technically different from a string saying "null" right now, and my intention here was just syntactic sugar - being able to use `timeout = null` rather than weird `timeout = "null"`. There's also no danger for this null being ambiguous (at least in terms of timeouts), because `"null"` is not a valid timeout in any case.

**Comment by avikivity** [comment]
Looks good. The only major comment is about having so many timeouts, but I guess Cassandra sets the expectations and it's natural to follow through.

File: cql3/Cql.g (hunk 1 above)
Comment (psarna) [reply cql3/Cql.g:1624]: ... and to back up my dubious claim that it's not a big issue (for the record - I also think it's rather dubious), unreserved keywords are treated pretty much the same way - so if somebody really wants to set `timeout = writetime`, which makes no sense, it would work just as good as `timeout = "writetime"`. My intention here was to treat null as unreserved in the particular case of property values.

File: db/system_distributed_keyspace.cc (hunk 3 above)
Comment (psarna) [reply db/system_distributed_keyspace.cc:181]: Original discussion:
https://github.com/scylladb/scylla/pull/7617#discussion_r533335847
https://github.com/scylladb/scylla/pull/7617#issuecomment-736792433
https://github.com/scylladb/scylla/pull/7617#issuecomment-737312128

tldr: we want a well-defined timestamp to avoid schema mismatches. Hence the previous patch which allows custom timestamps, and the C++ below to explicitly update with a well-defined timestamp (0).

**Comment by psarna** [comment]
v2:
 - dropped all 7 different timeout params in favor of a single
   `timeout` parameter, applicable to all operations; the resulting
   code is much clearer, tons of boilerplate is reduced and there's
   less room for mistakes
 - added a comment for updating tables with custom timestamps,
   explaining why the change is useful - both in code and in the
   commit message
 - removed a redundant co_return
 
Tests are updated accordingly and were rerun.

**Comment by avikivity** [comment]
> v2:
> 
> * dropped all 7 different timeout params in favor of a single
>   `timeout` parameter, applicable to all operations; the resulting
>   code is much clearer, tons of boilerplate is reduced and there's
>   less room for mistakes

I meant my comment to be the beginning of a conversation. But maybe it's for the best. If you issue a TRUNCATE from a session with a service level, you can override the timeout locally.

> * added a comment for updating tables with custom timestamps,
>   explaining why the change is useful - both in code and in the
>   commit message
> * removed a redundant co_return
> 
> Tests are updated accordingly and were rerun.

File: db/system_distributed_keyspace.cc (hunk 3 above)
Comment (avikivity) [reply db/system_distributed_keyspace.cc:181]: Aha. Perhaps we could have a clause for DDL to support this (but better to wait for Raft schemas).

**Comment by psarna** [comment]
> > v2:
> > 
> > * dropped all 7 different timeout params in favor of a single
> >   `timeout` parameter, applicable to all operations; the resulting
> >   code is much clearer, tons of boilerplate is reduced and there's
> >   less room for mistakes
> 
> I meant my comment to be the beginning of a conversation. But maybe it's for the best. If you issue a TRUNCATE from a session with a service level, you can override the timeout locally.

It was such a good beginning that I just immediately followed it with code to see how it comes out. I also took into account that Vlad has a similar opinion on multiple timeout values from scylla.yaml and the fact that I really hated typing all these timeouts so many times. Finally, if somebody is interested in finer granularity of timeouts, per-query timeouts are already out.

> 
> > * added a comment for updating tables with custom timestamps,
> >   explaining why the change is useful - both in code and in the
> >   commit message
> > * removed a redundant co_return
> > 
> > Tests are updated accordingly and were rerun.

File: cql3/Cql.g (hunk 1 above)
Comment (avikivity) [reply cql3/Cql.g:1624]: Ugh, the unreserved_keyword thing smells.

I worry that we'll sometimes fix it, then break unintentional code that quoted "null". Maybe use a string that's harder to get accidentally? e.g. "!?null?!" or something.

Comment (psarna) [reply cql3/Cql.g:1624]: Well, in case of timeouts, I can drop the patch that treats "null" as unreserved if we agree on another way for resetting a timeout value to default. Does `timeout = ""` sound acceptable? It doesn't sound too bad, although null sounded a little more standard.

Comment (avikivity) [reply cql3/Cql.g:1624]: No, NULL is the right thing here.

Comment (avikivity) [reply cql3/Cql.g:1624]: Also, please add a FIXME since the grammar should return a more complicated type (probably an expression that can be evaluated at compile type, or maybe just optional<string>.

Comment (psarna) [reply cql3/Cql.g:1624]: There's also a less intrusive way I came up with just now - instead of reusing propertyValue from the grammar, I can wrap it in a nullablePropertyValue, based on optionals, and use it just in the service level code.

Comment (avikivity) [reply cql3/Cql.g:1624]: Yes

Comment (psarna) [reply cql3/Cql.g:1624]: ... except is going to be really ugly, I just tried and gave up quickly. The whole `property_definitions` class is based on a map of strings, so it would either need to be templatized, or unconditionally use optionals underneath, or copy-pasted, and all of these options are a little too large to include them in this series for my taste - it's already a rather big change.

So instead, I suggest the following:
1. Add a FIXME in this series
2. Create an issue explaining why strings are not good enough to represent property values (they kind of are now, but they're not future-proof, because `null` is indistinguishable from `"null"` and same goes for all unreserved keywords).
3. Fix the above in a follow-up patch, outside of scope of this one. The CQL layer will remain identical from the user's point of view, but the underneath implementation would become more future-proof once (2) is fixed

**Comment by psarna** [comment]
v3:
 - added a fixme note mentioning that string is not a sufficient type to represent property values, especially after `null` is also allowed

File: cql3/Cql.g (hunk 1 above)
Comment (avikivity) [reply cql3/Cql.g:1629]: Ok.

Comment (psarna) [reply cql3/Cql.g:1629]: https://github.com/scylladb/scylla/issues/8598

Comment (nyh) [reply cql3/Cql.g:1629]: @avikivity @psarna what about `with timeout = infinity`? If I read the parser correctly, it accepts the keyword `infinity` as an infinite (floating point) constant so it sounds a reasonable choice, and it doesn't require this `unreserved_keyword` thing (maybe we'll even be lucky and can remove this `unreserved_keyword` production from `property`?).

File: service/qos/qos_common.hh [hunk 11]
```diff
@@ -36,8 +36,12 @@ namespace qos {
  *  a service level.
  */
 struct service_level_options {
+    static constexpr lowres_clock::duration delete_marker = lowres_clock::duration::max();
+
```
Comment (avikivity) [inline service/qos/qos_common.hh:40]: How is delete_marker different from timeout == nullopt, below?

Comment (avikivity) [reply service/qos/qos_common.hh:40]: If you need three options, how about variant<unspecified, deleted, lowres_clock::duration>?

File: cql3/Cql.g (hunk 1 above)
Comment (psarna) [reply cql3/Cql.g:1629]: Using infinity sounds like a nice potential extension, but for a different purpose. `null` is about deleting the per-service-level timeout and defaulting back to the one defined in `scylla.yaml`. Setting it to infinity sounds more like actually setting the operation's timeout to "never time this operation out".

Comment (psarna) [reply cql3/Cql.g:1629]: (and the latter can currently be worked around quite neatly by setting the timeout to an unbelievably high value)

File: service/qos/qos_common.hh (hunk 11 above)
Comment (psarna) [reply service/qos/qos_common.hh:40]: Well, now I don't actually need this marker at all, the code can be simplified even further. Originally, the purpose was to distinguish between setting `read_timeout = null and write_timeout = 5s` and `write_timeout = 5s`. The former would actively delete the per-service-level timeout value, the latter would simply not touch it and only set the write timeout. Since I kept all the timeout values in a single struct, I needed to represent both "please leave untouched" and "please delete" somehow. Now that there's only 1 timeout value, I no longer need this dance. Perhaps it will be useful to introduce it later when more service level options are added, but there's no point in complicating the code now.

Comment (psarna) [reply service/qos/qos_common.hh:40]: Hm, but on the other hand I **do** plan to add more per-service-level options soon. So perhaps I'll indeed make it an explicit variant in v3.

Comment (avikivity) [reply service/qos/qos_common.hh:40]: Yes, there's a difference between a mutation to service level options (which can delete a setting) and s value of service level options (which can represent a setting that was deleted or never set, but not the deletion itself).

**Comment by psarna** [comment]
v4:
 - migrated the timeout parameter from being represented as  `std::optional + a special value` to `std::variant<unset, plz_delete_me, actual_value>`, in order to be more explicit; as a result most of the related code was rewritten with the visitor pattern

File: docs/service_levels.md [hunk 12]
```diff
@@ -33,6 +33,69 @@ one can run the following query:
...
+         role2
+          /
+       role1
+        
+With the following roles granted, here are the effective timeouts for the roles:
+
+role1: `timeout = 10ms`
+role2: `timeout = 10ms`
+role3: `timeout = 2s`
+role4: `timeout = 10ms`
```
Comment (avikivity) [inline docs/service_levels.md:100]: We can consider adding a priority field to service levels, so that if someone has multiple service levels attached then the one with the highest priority is chosen. At this point it's a needless complication though.

File: service/qos/qos_common.cc [hunk 13]
```diff
@@ -20,5 +20,25 @@
...
+            ret.timeout = default_values.timeout;
+        },
+        [&] (const delete_marker& dm) {
+            // remove the value
+            ret.timeout = unset_marker{};
+        },
+        [&] (const lowres_clock::duration&) {
+            // leave the value as is
+        },
+    }, ret.timeout);
```
Comment (avikivity) [inline service/qos/qos_common.cc:40]: This is nice.

File: docs/service_levels.md (hunk 12 above)
Comment (psarna) [reply docs/service_levels.md:100]: That, or every parameter can have its own independent policy for reducing all values into one - e.g. with the future oltp/olap parameter it would make sense to take the value from the most immediate ancestor that can be found or warn if contradicting records are found. In any case, I left it for future consideration.
